lib/cchain/provider/provider.go

```diff
@@ -29,7 +29,10 @@ import (
 
 var _ cchain.Provider = Provider{}
 
-type fetchFunc func(ctx context.Context, chainVer xchain.ChainVersion, fromOffset uint64) ([]xchain.Attestation, error)
+// fetchFunc fetches all attestations at offset for provided chain version. It returns found
+// attestations as well as the consensus block height at which attestations were found.
+// If offset is found at the latest block attestations are returned but height returned is 0.
+type fetchFunc func(ctx context.Context, chainVer xchain.ChainVersion, fromOffset uint64, cachedHeight uint64) ([]xchain.Attestation, uint64, error)
```

Comment:
I'd suggest (nit, rename `cachedHeight` to `searchStartHeight`):
```
fetchFunc returns a slice of strictly-sequential attestations for the provided fromOffset (inclusive) and chain version. It also returns the consensus block height where these attestations were found. This height can be provided in subsequent calls as searchStartHeight to improve performance by skipping the internal lookbackRange step. Providing a zero searchStartHeigt first performs an exponential lookback to identify binary search range.

If the consensus chain doesn't contain the fromOffset yet, the latest height is returned since this is safe to use as searchStartHeight. 
```

Reply:
Added a comment. The variable name is better. One nit feedback tho is that I believe comments explaining interfaces shouldn't contain implementation details. In this case I believe explaining about lookbackRange etc is an implementation detail, and since interfaces are APIs we shouldn't explain how things get implemented. I think the comment could end at: searchStartHeight is used to narrow the height search range.